aten/src/ATen/native/FractionalMaxPool2d.cpp

```diff
@@ -0,0 +1,393 @@
+    /* indices will contain the locations for each output point */
+    indices.resize_({numBatch, numPlanes, outputH, outputW});
+  }
+
+  AT_DISPATCH_FLOATING_TYPES(input.type(),
+  "fractional_max_pool2d_out_frame", [&] {
+    auto input_data = input.data<scalar_t>();
+    auto output_data = output.data<scalar_t>();
+    auto indices_data = indices.data<int64_t>();
+    auto randomSamples_data = randomSamples.data<scalar_t>();
+    fractional_max_pool2d_out_frame<scalar_t>(
+      input_data,
+      output_data,
+      indices_data,
+      randomSamples_data,
+      numBatch, numPlanes,
+      inputW, inputH,
+      outputW, outputH,
+      poolSizeW, poolSizeH);
+    }
+  );
+}
+
+template <typename scalar_t>
+static void fractional_max_pool2d_backward_out_single_batch_frame(
+  scalar_t* gradInput,
+  scalar_t* gradOutput,
+  int64_t* indices,
+  int numPlanes,
+  int inputW, int inputH,
+  int outputW, int outputH) {
+  int plane;
+#pragma omp parallel for private(plane)
+  for (plane = 0; plane < numPlanes; plane++) {
+    scalar_t* gradInputForPlane = gradInput + plane * inputW * inputH;
+    scalar_t* gradOutputForPlane = gradOutput + plane * outputW * outputH;
+    int64_t* indicesForPlane = indices + plane * outputW * outputH;
+
+    int h, w;
+    for (h = 0; h < outputH; ++h) {
+      for (w = 0; w < outputW; ++w) {
+        int outputIndex = h * outputW + w;
+        int64_t index = indicesForPlane[outputIndex];
+        AT_ASSERT(index >= 0 && index < inputW * inputH);
+
+        gradInputForPlane[index] += gradOutputForPlane[outputIndex];
+      }
+    }
+  }
+}
+
+template <typename scalar_t>
+static void fractional_max_pool2d_backward_out_frame(
+  scalar_t* gradInput,
+  scalar_t* gradOutput,
+  int64_t* indices,
+  int numBatch, int numPlanes,
+  int inputW, int inputH,
+  int outputW, int outputH) {
+    if(numBatch == 1) {
+      fractional_max_pool2d_backward_out_single_batch_frame<scalar_t>(
+        gradInput, gradOutput, indices,
+        numPlanes,
+        inputW, inputH, outputW, outputH
+      );
+      return;
+    }
+    int batch;
+#pragma omp parallel for private(batch)
+    for(batch = 0; batch < numBatch; ++ batch) {
+      fractional_max_pool2d_backward_out_single_batch_frame<scalar_t>(
+        gradInput + batch * numPlanes * inputH * inputW,
+        gradOutput + batch * numPlanes * outputH * outputW,
+        indices + batch * numPlanes * outputH * outputW,
+        numPlanes, inputW, inputH, outputW, outputH);
+    }
+}
+
+Tensor& fractional_max_pool2d_backward_out_cpu_template(
+  const at::Tensor& input,
+  const at::Tensor& gradOutput_,
+  at::Tensor& gradInput,
+  IntList output_size,
+  IntList pool_size,
+  const at::Tensor& indices) {
+
+  int numBatch = 1;
+  int planeDim = 0;
+  int heightDim = 1;
+  int widthDim = 2;
+
+  int outputW = output_size[0];
+  int outputH = output_size[1];
+  int poolSizeW = pool_size[0];
+  int poolSizeH = pool_size[1];
```

Comment:
Unfortunately I gave bad guidance here. I'm sorry but I gave you bad guidance here. `output_size[0]` is height, `output_size[1]` is width. The THNN ordering here is not the right thing to respect; instead we need to respect the documentation. I wrote some more guidance in the porting guide, which I've reproduced below:

## Height-width argument order

Some kernels in THNN have very strange argument order; for example:

```
void THNN_(SpatialFractionalMaxPooling_updateOutput)(
           THCState *state,
           THCTensor *input,
           THCTensor *output,
           int outputW, int outputH,
           int poolSizeW, int poolSizeH,
           THCIndexTensor *indices,
           THCTensor *randomSamples)
```

outputW comes before outputH, even though conventional NCHW ordering says that height comes before width. *There is code generation logic specifically for THNN that reorders output_sizes to match the THNN kernel; you should manually reorder when you port.* So for example, the above arguments become:

```
void fractional_max_pool2d_out(
  Tensor& output,
  Tensor& indices,
  const Tensor& input,
  IntList pool_size,
  IntList output_size,
  const Tensor& randomSamples) {
  ...
  auto outputH = output[0]; // Notice they are reordered
  auto outputW = output[1];
  auto poolSizeH = pool_size[0];
  auto poolSizeW = pool_size[1];
}
```

The tests are supposed to catch if you've gotten it wrong, but we may have missed some cases, so be vigilant. In general, height comes before width. Here is a full list of dimension offsets (from aten/src/ATen/nn_parse.py), where negative numbers mean that you index from the end of the list:

```
DIMENSION_OFFSET = {
    'width': -1,
    'height': -2,
    'B': 0,
    'C': 1,
    'W': -1,
    'H': -2,
    'T': -3,
    'left': 0,
    'right': 1,
    'top': 2,
    'bottom': 3,
    'front': 4,
    'back': 5,
}
```